source/server/listener_impl.cc

```diff
@@ -82,100 +82,121 @@ bool shouldBindToPort(const envoy::config::listener::v3::Listener& config) {
-  ENVOY_LOG(debug, "Create listen socket for listener {} on address {}", listener_name_,
+  ENVOY_LOG(debug, "Create listen socket for listener {} on address {}", listener_name,
             local_address_->asString());
-  if (socket != nullptr && options_ != nullptr) {
+  if (socket != nullptr && options != nullptr) {
     const bool ok = Network::Socket::applyOptions(
-        options_, *socket, envoy::config::core::v3::SocketOption::STATE_BOUND);
+        options, *socket, envoy::config::core::v3::SocketOption::STATE_BOUND);
     const std::string message =
-        fmt::format("{}: Setting socket options {}", listener_name_, ok ? "succeeded" : "failed");
+        fmt::format("{}: Setting socket options {}", listener_name, ok ? "succeeded" : "failed");
     if (!ok) {
       ENVOY_LOG(warn, "{}", message);
-      throw Network::CreateListenerException(message);
+      throw Network::SocketOptionException(message);
     } else {
       ENVOY_LOG(debug, "{}", message);
     }
 
     // Add the options to the socket_ so that STATE_LISTENING options can be
-    // set in the worker after listen()/evconnlistener_new() is called.
-    socket->addOptions(options_);
+    // set after listen() is called and immediately before the workers start running.
+    socket->addOptions(options);
   }
   return socket;
 }
 
-Network::SocketSharedPtr ListenSocketFactoryImpl::getListenSocket() {
-  if (!reuse_port_) {
-    // We want to maintain the invariance that listeners do not share the same
-    // underlying socket. For that reason we return a socket based on a duplicated
-    // file descriptor.
-    return socket_->duplicate();
-  }
-
-  Network::SocketSharedPtr socket;
-  absl::call_once(steal_once_, [this, &socket]() {
-    if (socket_) {
-      // If a listener's port is set to 0, socket_ should be created for reserving a port
-      // number, it is handed over to the first worker thread came here.
-      // There are several reasons for doing this:
-      // - for UDP, once a socket being bound, it begins to receive packets, it can't be
-      //   left unused, and closing it will lost packets received by it.
-      // - port number should be reserved before adding listener to active_listeners_ list,
-      //   otherwise admin API /listeners might return 0 as listener's port.
-      socket = std::move(socket_);
-    }
-  });
+Network::SocketSharedPtr ListenSocketFactoryImpl::getListenSocket(uint32_t socket_index) {
+  ASSERT(socket_index < sockets_.size() && sockets_[socket_index] != nullptr);
+  return sockets_[socket_index];
+}
 
-  if (socket) {
-    return socket;
+void ListenSocketFactoryImpl::doFinalPreWorkerInit() {
+  if (bind_type_ == ListenerComponentFactory::BindType::NoBind ||
+      socket_type_ != Network::Socket::Type::Stream) {
+    return;
   }
 
-  return createListenSocketAndApplyOptions();
+  for (auto& socket : sockets_) {
+    // TODO(mattklein123): At some point we lost error handling on this call which I think can
```

Comment:
Sure I can do that, but if I do that I might as well just add the real error handling in this PR?

Reply:
I think I'm going to leave the TODO on this one just because the PR is so large. I will do a fast follow with adding error handling for the 2 listen calls.